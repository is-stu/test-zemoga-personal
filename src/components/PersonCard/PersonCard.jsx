import { Button, Card, Col, ProgressBar, Row, Spinner } from 'react-bootstrap'
import { ThumbsDown, ThumbsUp } from 'akar-icons'
import moment from 'moment/moment'

const iconByVows = (person) => {
  return person?.votes?.positive > person?.votes?.negative
    ? (
      <ThumbsUp strokeWidth={2} size={20} color='green' className='me-1' />
      )
    : (
      <ThumbsDown strokeWidth={2} size={20} color='red' className='me-1' />
      )
}

export const PersonCard = ({ person, updatePerson }) => {
  const positivePercentage =
    (person?.votes?.positive /
      (person?.votes?.positive + person?.votes?.negative)) *
    100
  const negativePercentage =
    (person?.votes?.negative /
      (person?.votes?.positive + person?.votes?.negative)) *
    100

  const handleThumbsUp = () => {
    const payload = {
      ...person,
      votes: {
        positive: person.votes.positive + 1,
        negative: person.votes.negative
      }
    }
    updatePerson.execute(payload, {
      params: {
        id: person?._id
      }
    })
  }

  const handleThumbsDown = () => {
    const payload = {
      ...person,
      votes: {
        positive: person.votes.positive,
        negative: person.votes.negative + 1
      }
    }
    updatePerson.execute(payload, {
      params: {
        id: person?._id
      }
    })
  }

  return (
    <>
      <Card
        className='mb-1'
        style={{
          width: '40rem',
          border:
            person?.votes?.positive > person?.votes?.negative
              ? '2px solid green'
              : '2px solid red'
        }}
      >
        <Card.Img
          variant='top'
          src={person?.picture}
          style={{ maxHeight: '400px', objectFit: 'cover' }}
        />
        <Card.Body>
          <Card.Title>
            {iconByVows(person)}
            {person?.name}
          </Card.Title>
          <Card.Subtitle>
            {moment(person?.lastUpdated).fromNow()} in {person?.category}
          </Card.Subtitle>
          <Card.Text>{person?.description}</Card.Text>
          <Card.Text className='d-flex justify-content-center fw-bold'>
            What's your veredict ?
          </Card.Text>
          <Row>
            <Col className='d-flex justify-content-center'>
              <Button variant='outline-success' onClick={handleThumbsUp}>
                {updatePerson.loading
                  ? (
                    <Spinner />
                    )
                  : (
                    <ThumbsUp strokeWidth={2} size={36} />
                    )}
              </Button>
            </Col>
            <Col className='d-flex justify-content-center'>
              <Button variant='outline-danger' onClick={handleThumbsDown}>
                {updatePerson.loading
                  ? (
                    <Spinner />
                    )
                  : (
                    <ThumbsDown strokeWidth={2} size={36} />
                    )}
              </Button>
            </Col>
          </Row>
          <ProgressBar className='mt-3'>
            <ProgressBar
              variant='success'
              now={positivePercentage}
              key={1}
              label={`${positivePercentage.toFixed(1)} %`}
            />
            <ProgressBar
              variant='danger'
              now={negativePercentage}
              key={2}
              label={`${negativePercentage.toFixed(1)} %`}
            />
          </ProgressBar>
        </Card.Body>
      </Card>
    </>
  )
}
