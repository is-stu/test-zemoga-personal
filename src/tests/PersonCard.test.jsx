/* eslint-env jest */

import { render } from '@testing-library/react'
import { PersonCard } from '../components/PersonCard/PersonCard'

describe('PersonCard Testing', () => {
  test('should match with snapshot', () => {
    const { container } = render(<PersonCard />)

    expect(container).toMatchSnapshot()
  })
})
