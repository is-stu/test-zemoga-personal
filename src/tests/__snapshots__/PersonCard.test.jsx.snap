// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonCard Testing should match with snapshot 1`] = `
<div>
  <div
    class="mb-1 card"
    style="width: 40rem; border: 2px solid red;"
  >
    <img
      class="card-img-top"
      style="max-height: 400px; object-fit: cover;"
    />
    <div
      class="card-body"
    >
      <div
        class="card-title h5"
      >
        <svg
          class="me-1"
          fill="none"
          height="20"
          stroke="red"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.395 18.218l-.23 2.369c-.091.952-.98 1.598-1.878 1.366-1.351-.35-2.3-1.605-2.3-3.044v-3.035c0-.675 0-1.013-.146-1.26a1.018 1.018 0 0 0-.333-.345c-.24-.151-.567-.151-1.22-.151h-.396c-1.703 0-2.554 0-3.078-.39a2.073 2.073 0 0 1-.78-1.208c-.146-.65.181-1.463.836-3.087l.327-.81c.188-.468.265-.975.225-1.48-.232-2.874 2.047-5.295 4.833-5.135l10.424.598c1.139.065 1.708.098 2.222.553.515.455.612.924.805 1.861a14.317 14.317 0 0 1-.055 6.037c-.283 1.248-1.475 1.92-2.706 1.76-3.264-.42-6.223 2.019-6.55 5.4z"
          />
          <path
            d="M17 12.5l.137-.457c.887-2.956.84-6.115-.137-9.043"
          />
        </svg>
      </div>
      <div
        class="card-subtitle h6"
      >
        a few seconds ago
         in 
      </div>
      <p
        class="card-text"
      />
      <p
        class="d-flex justify-content-center fw-bold card-text"
      >
        What's your veredict ?
      </p>
      <div
        class="row"
      >
        <div
          class="d-flex justify-content-center col"
        >
          <button
            class="btn btn-outline-success"
            type="button"
          >
            <svg
              fill="none"
              height="36"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="36"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.605 5.782l.23-2.369c.091-.952.98-1.598 1.878-1.366 1.351.35 2.3 1.605 2.3 3.044v3.035c0 .675 0 1.013.146 1.26.083.141.197.26.333.345.24.151.567.151 1.22.151h.396c1.703 0 2.554 0 3.078.39.393.293.67.722.78 1.208.146.65-.181 1.463-.836 3.087l-.326.81a3.261 3.261 0 0 0-.226 1.48c.232 2.874-2.047 5.295-4.833 5.136l-10.424-.599c-1.139-.065-1.708-.098-2.222-.553-.515-.455-.612-.924-.805-1.861a14.324 14.324 0 0 1 .055-6.037c.283-1.248 1.475-1.92 2.706-1.76 3.264.42 6.223-2.019 6.55-5.4z"
              />
              <path
                d="M7 11.5l-.137.457A14.983 14.983 0 0 0 7 21"
              />
            </svg>
          </button>
        </div>
        <div
          class="d-flex justify-content-center col"
        >
          <button
            class="btn btn-outline-danger"
            type="button"
          >
            <svg
              fill="none"
              height="36"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="36"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12.395 18.218l-.23 2.369c-.091.952-.98 1.598-1.878 1.366-1.351-.35-2.3-1.605-2.3-3.044v-3.035c0-.675 0-1.013-.146-1.26a1.018 1.018 0 0 0-.333-.345c-.24-.151-.567-.151-1.22-.151h-.396c-1.703 0-2.554 0-3.078-.39a2.073 2.073 0 0 1-.78-1.208c-.146-.65.181-1.463.836-3.087l.327-.81c.188-.468.265-.975.225-1.48-.232-2.874 2.047-5.295 4.833-5.135l10.424.598c1.139.065 1.708.098 2.222.553.515.455.612.924.805 1.861a14.317 14.317 0 0 1-.055 6.037c-.283 1.248-1.475 1.92-2.706 1.76-3.264-.42-6.223 2.019-6.55 5.4z"
              />
              <path
                d="M17 12.5l.137-.457c.887-2.956.84-6.115-.137-9.043"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="mt-3 progress"
      >
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="NaN"
          class="progress-bar bg-success"
          role="progressbar"
        >
          NaN %
        </div>
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="NaN"
          class="progress-bar bg-danger"
          role="progressbar"
        >
          NaN %
        </div>
      </div>
    </div>
  </div>
</div>
`;
